pkg/vsphere/vsan/synced_cache.go

```diff
@@ -0,0 +1,439 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package vsan is a workaround for vsan DOM object leaking issue caused by FileManager.DeleteDatastoreFile, see github issue #3787 and bugzilla issue #1808703
+// This file used draft vSphere API, which is subject to change in the future, so this workaround should be removed as soon as the DOM leaking issue is fixed by vSAN.
+package vsan
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/pkg/errors"
+	"github.com/vmware/vic/pkg/trace"
+)
+
+var (
+	SyncedDomCache  = newDomCache()
+	refreshInterval = 5 * time.Minute
+)
+
+func init() {
+	go SyncedDomCache.run()
+}
+
+type DomDeleteError struct {
+	Err         error
+	FailedUuids []string
+	// if delete dom objects returns no error, but with some of uuid delete failure, this result will return detail error information for each uuid
+	Result []types.HostVsanInternalSystemDeleteVsanObjectsResult
+}
+
+func (e DomDeleteError) Error() string {
+	if e.Err != nil {
+		return e.Err.Error()
+	}
+	return fmt.Sprintf("Failed to delete uuids: %s", e.FailedUuids)
+}
+
+type DSDomCache interface {
+	Refresh(ctx context.Context) error
+	DeleteVMDKDoms(ctx context.Context, paths []string) ([]string, error)
+	CleanOrphanDoms(ctx context.Context) ([]string, error)
+}
+
+type dsWaitGroup struct {
+	dsRef string
+	wg    *sync.WaitGroup
+}
+
+type syncedDomCache struct {
+	// datastore map - this cache can support multiple vsan datastores
+	dsMap map[string]DSDomCache
+	m     *sync.RWMutex
+	// refresh request channel, dsWaitGroup.wg.Done() will be called after this datastore's refresh is done
+	refresh chan *dsWaitGroup
+	// channel for stop request
+	stop chan bool
+	// finished refresh channel, after this request is accepted, all waitgroups for this datastore refresh will be invoked
+	doneOnce chan string
+	// wait refresh request channel, which does not trigger new refresh, just wait for current one if there is
+	waitCurrentRefresh chan *dsWaitGroup
+	// map to hold all waitgroups for one datastore refresh
+	refreshings map[string][]*sync.WaitGroup
+}
+
+func newDomCache() *syncedDomCache {
+	return &syncedDomCache{
+		dsMap:              make(map[string]DSDomCache),
+		m:                  &sync.RWMutex{},
+		refresh:            make(chan *dsWaitGroup),
+		stop:               make(chan bool, 1),
+		doneOnce:           make(chan string),
```

Comment:
This probably needs to be buffered since we do write to this channel multiple values at a time.